### PR TITLE
WIP: Allow admission controllers from Kubernetes to be used in Service Catalog apiserver

### PR DESCRIPTION
**Pull request by ychen-atlassian**
### Background

Currently Service Catalog has built its own various admission controllers in the `plugins/pkg/admission` package.

While some of the Kubernetes Admission Controllers are not always useful for service-catalog (e.g. AlwaysPullImages, PodPreset, etc) there are a few useful ones which still can still work with service-catalog, e.g. ValidatingAdmissionWebhook, MutatingAdmissionWebhook, AlwaysAdmit, AlwaysDeny, EventRateLimit, OwnerReferencesPermissionEnforcement, etc.

### Changes

- Service Catalog's admission initializer originally attempted to call `WantsKubeClientSet` and `WantsKubeInformerFactory`, while Kubernetes initializer called `initializers.WantsExternalKubeClientSet` and `initializers.WantsExternalKubeClientSet`. This code has been reconciled and references the k8s.io types.
- We also now call `initializers.WantsScheme` and `initializers.WantsAuthorizer` to initialise other stuff Kubernetes admission controllers want.
- Install the types in `github.com/kubernetes-incubator/service-catalog/pkg/kubernetes/pkg/apis/admission/install` (which is just AdmissionReview at the moment?) and `github.com/kubernetes-incubator/service-catalog/pkg/kubernetes/pkg/apis/authentication/install`, which is required by the AdmissionReview.
- Because `k8s.io/kubernetes/pkg/apis/authentication|admission/install` hasn't been moved out to the k8s.io api yet, the files have been copied into the pkg/kubernetes directory
- The `KubernetesNamespaceLifecycle` admission controller in Service Catalog now registers the existing `NamespaceLifecycle` controller from k8s.io under the existing Service Catalog admission controller name, for backwards compatibility reasons.
  - (Actually regarding this - it might be worth changing the deployment yaml to use NamespaceLifecycle and to eventually get rid of KubernetesNamespaceLifecycle)

### Other

- Size should probably just be size/S, because most of the files here are from copying Kubernete code over.
- This has been manually tested with ValidatingAdmissionWebhook, and it does end up working.

### CODE REVIEW

File: cmd/apiserver/app/server/util.go [hunk 1]
```diff
@@ -148,7 +149,7 @@ func buildGenericConfig(s *ServiceCatalogServerOptions) (*genericapiserver.Recom
 
 		// TODO: we need upstream to package AlwaysAdmit, or stop defaulting to it!
```
Comment (nilebox) [inline cmd/apiserver/app/server/util.go]: @derekwaynecarr @pmorie unrelated note: is this TODO still valid? I can't find "AlwaysAdmit" anywhere in the code.
Asking because in this PR there is an authorizer introduced that might affect this TODO?

File: pkg/api/register.go [hunk 2]
```diff
@@ -43,6 +45,8 @@ var (
 func init() {
 	servicecataloginstall.Install(groupFactoryRegistry, Registry, Scheme)
 	settingsinstall.Install(groupFactoryRegistry, Registry, Scheme)
+	admissioninstall.Install(groupFactoryRegistry, Registry, Scheme)
```
Comment (nilebox) [inline pkg/api/register.go]: Given that this is not specific to Service Catalog (or Kubernetes API server), we should eventually move this generic admission-related code from the main `k8s.io/kubernetes` repo to `k8s.io/apiserver` and reuse it.
But since Service Catalog doesn't have a dependency on `k8s.io/kubernetes` anymore, it should be fine to copy the code into `pkg/kubernetes/...` for now.

File: pkg/apiserver/admission/initializer.go [hunk 3]
```diff
@@ -38,36 +41,34 @@ type WantsInternalServiceCatalogInformerFactory interface {
 	admission.InitializationValidator
 }
 
-// WantsKubeClientSet defines a function which sets ClientSet for admission plugins that need it
-type WantsKubeClientSet interface {
```
Comment (nilebox) [inline pkg/apiserver/admission/initializer.go]: This change was most likely already made upstream, we just haven't adopted this change when rebased on the newer releases of Kubernetes.

File: plugin/pkg/admission/namespace/lifecycle/admission.go [hunk 4]
```diff
@@ -17,147 +17,35 @@ limitations under the License.
...
-	// 1. It allows a watch-fed cache time to observe a namespace creation event
-	// 2. It allows time for a namespace creation to distribute to members of a storage cluster,
-	//    so the live lookup has a better chance of succeeding even if it isn't performed against the leader.
-	missingNamespaceWait = 50 * time.Millisecond
 )
 
 // Register registers a plugin
 func Register(plugins *admission.Plugins) {
+	// This essentially registers the existing NamespaceLifecycle for kubernetes
+	// another another name, for backwards-compatibility reasons
```
Comment (nilebox) [inline plugin/pkg/admission/namespace/lifecycle/admission.go]: "another another" -> please fix the comment.

File: plugin/pkg/admission/broker/authsarcheck/admission_test.go [hunk 5]
```diff
@@ -43,7 +43,7 @@ func newHandlerForTest(kubeClient kubeclientset.Interface) (admission.Interface,
 	if err != nil {
 		return nil, kf, err
 	}
-	pluginInitializer := scadmission.NewPluginInitializer(nil, nil, kubeClient, kf)
+	pluginInitializer := scadmission.NewPluginInitializer(nil, nil, kubeClient, kf, nil, nil)
```
Comment (nilebox) [inline plugin/pkg/admission/broker/authsarcheck/admission_test.go]: We don't seem to have any integration tests to make sure that these new arguments and changes in pluginInitializer don't break anything.
@jpeeler can you review please?

**Comment by nilebox** [comment]
The Jenkins build fails on installing ups-broker using helm.
I think I have seen this failed before:
```
2. Executing helm install /var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/charts/ups-broker --name ups-broker --namespace ups-broker --set image=gcr.io/service-catalog-jenkins/catalog/user-broker:75a4e5b
Error: release ups-broker failed: services "ups-broker-ups-broker" already exists
Command failed. Will retry in 7 seconds.
```
It's either a temporary issue, or some flakiness.

**Comment by nilebox** [comment]
@kibbles-n-bytes can you take a look at the failure?
It's this block [test_walkthrough.sh#L86-L91](https://github.com/kubernetes-incubator/service-catalog/blob/master/contrib/jenkins/test_walkthrough.sh#L86-L91):
```
retry \
    helm install "${ROOT}/charts/ups-broker" \
    --name "ups-broker" \
    --namespace "ups-broker" \
    --set "${VALUES}" \
  || error_exit 'Error deploying ups-broker to cluster.'
```
Doesn't seem to be related to the changes in this PR.

**Comment by kibbles-n-bytes** [comment]
Looks to have been a transient error; strange. Everything's green now.

**Comment by MHBauer** [comment]
I don't know what most of these admission controllers do, so I don't know what we're gaining. Why do we need this?

Does anybody from apimachinery have an opinion on this? @deads2k @sttts @liggitt 
I'd prefer to have upstream support rather than copying in a bunch of source.

Do we end up running some of these twice when aggregation is enabled?
If we do have aggregation enabled, wouldn't we already have hit any of these installed at the cluster level?

**Comment by nilebox** [comment]
> I don't know what most of these admission controllers do, so I don't know what we're gaining. Why do we need this?

We at Atlassian want to perform some additional mutation and validation, and it's enabled through `MutatingAdmissionWebhook` and `ValidatingAdmissionWebhook` admission controllers correspondingly. Traditionally, with "static admission controllers", we would have to fork the Service Catalog repo and built our own API server with custom admission controllers from source.
See https://kubernetes.io/docs/admin/extensible-admission-controllers/#external-admission-webhooks for more details

> Does anybody from apimachinery have an opinion on this?
> I'd prefer to have upstream support rather than copying in a bunch of source.

@MHBauer yes, we are working on the upstream support in parallel, but it will take some time.
I will add a `do-not-merge` label to this PR until we discuss it.
@deads2k commented yesterday that webhooks should be supported by custom API servers since k8s 1.9, but in our experience it actually requires more work to enable Validation and Mutation webhooks and using the code that lives in the main `k8s.io/kubernetes` repo. 
We plan to prepare an example with https://github.com/kubernetes/sample-apiserver to demonstrate what is exactly missing in the upstream.

> If we do have aggregation enabled, wouldn't we already have hit any of these installed at the cluster level?

Unfortunately not, admission controllers are specific to each API server, and there are no "global" admission controllers in the aggregation server.
If you are interested in more details, see https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/apiserver.go
The core set of API servers consists of 3 servers:
- Core API server (managing Pods, Deployments, Secrets, ConfigMaps, etc) - has admission controllers registered including webhooks https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/server.go#L519
- Extensions API server (CRDs) - supports admission webhooks as well
- Aggregator API server (proxy for auth + routing requests to other API server) - doesn't have any admission controllers https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/aggregator.go

@ychen-atlassian feel free to add more details or correct me if some statements are wrong.

**Comment by ychen-atlassian** [comment]
W.R.T. aggregation layer - specifically while all 3 servers *do* end up reading the `--admission-control` flag and initializing the listed admission plugins, the aggregation server does not actually do anything regarding AdmissionControl:

Code:

1. In kubernetes/cmd/kube-apiserver/app/server.go, we apply the AdmissionOptions to server Config. This sets the `AdmissionControl` field in `server.Config` as well as running some init code on the admission controllers. The init code does nothing more than set some variables. The `server.Config` is eventually passed to the _apiserver_, _apiserver-extensions_ and _aggregator_ servers. TL;DR: we know if a server is doing anything with Admission controllers, it will need to do something with `c.AdmissionControl`
2. The _apiserver_ and _apiserver-extensions_ calls `New()` using the kubernetes apiserver `k8s.io/apiserver`, which does a bunch of stuff and creates a `GenericApiServer` with the right things created and all the special sauce to make AdmissionControllers work.
3. _aggregator_ does not use `AdmissionControl` - it ends up using it's own special code in `k8s.io/kube-aggregator/pkg/apiserver`. Grepping through the code produces nothing of relevance to admission controllers.
4. TL;DR: aggregator does not do anything special, it just aggregates. Service catalog will need to do something similar to apiserver and apiserver-extensions

You can test this by just deploying service catalog on a kubernetes cluster and enabling `--admission-control=AlwaysFail`, and creating a ServiceInstance and the operation will still succeed.

File: charts/catalog/templates/apiserver-deployment.yaml [hunk 6]
```diff
@@ -40,7 +40,7 @@ spec:
         - {{ .Values.apiserver.audit.logPath }}
         {{- end}}
         - --admission-control
```
Comment (sttts) [inline charts/catalog/templates/apiserver-deployment.yaml]: is this already kube 1.10 based? `--admission-control` will be deprectated in favor of `--enable-admission-plugins` to selectively enable plugins. The `AdmissionOptions` in 1.10 has a `RecommendedPluginOrder` selection of plugins, pre-set with the recommended list for aggregated apiservers.

File: cmd/apiserver/app/server/util.go (hunk 1 above)
Comment (sttts) [reply cmd/apiserver/app/server/util.go]: AlwaysAdmit is in k8s.io/apiserver, but not used by default (anymore?)

**Comment by sttts** [comment]
> commented yesterday that webhooks should be supported by custom API servers since k8s 1.9, but in our experience it actually requires more work to enable Validation and Mutation webhooks and using the code that lives in the main k8s.io/kubernetes repo.

This was a bug, fixed in https://github.com/kubernetes/kubernetes/pull/60965.

About @ychen-atlassian's comment:

> 1) Conclusion: apiservers must reference the AdmissionControl field in the Config to enable admission controllers.

This will be automatic in 1.10: https://github.com/kubernetes/kubernetes/blob/201c11f147c85b029665915bee3a62eea19d6d57/staging/src/k8s.io/apiserver/pkg/server/options/recommended.go#L64

> 2) Conclusion: apiserver and apiserver-extensions does something with admission controllers.

Admission is wired for resources served by the server, e.g. APIServices for the aggregator. But admission is not active on any other HTTP endpoint like the proxy endpoints of the aggregator.

> 3) aggregator does not use AdmissionControl - it ends up using it's own special apiserver code in k8s.io/kube-aggregator/pkg/apiserver

Wrong. It does use admission for APIServices.

> TL;DR: aggregator does not do anything special, it just aggregates. Service catalog will need to do something similar to apiserver and apiserver-extensions.

I would propose: rebase to 1.10 and use the `RecommendedOptions`. Add your own admission controllers to the recommended list in the `AdmissionOptions`. Remove `--admission-control` from the helm templates.

File: charts/catalog/templates/apiserver-deployment.yaml (hunk 6 above)
Comment (nilebox) [reply charts/catalog/templates/apiserver-deployment.yaml]: @sttts No, Service Catalog is currently based on 1.9, but thanks for the useful information, I will definitely try to migrate it on 1.10 soon.

**Comment by MHBauer** [comment]
This seems pretty huge. Every different apiserver has different admission controllers, and need to be manually synced up or manually configured each time. Even if we did provide this, how do we make sure people know about it to use it?

**Comment by sttts** [comment]
@MHBauer compare https://github.com/kubernetes-incubator/service-catalog/pull/1811#discussion_r173403788. We changed all the admission logic exactly for that reason. No admin wants to configure that each time. You as an apiserver producer can do it in the code once and for all.

File: charts/catalog/templates/apiserver-deployment.yaml (hunk 6 above)
Comment (MHBauer) [reply charts/catalog/templates/apiserver-deployment.yaml]: @sttts thanks for the pointer. good to have defaults. Is the underlying 1.10 lib of apistuff tagged and released?

Comment (sttts) [reply charts/catalog/templates/apiserver-deployment.yaml]: Yes, tagged with the usual automatic tags: https://github.com/kubernetes/apiserver/releases/tag/kubernetes-1.10.0-beta.2

**Comment by ychen-atlassian** [comment]
Sorry @sttts, I think I made things a bit complicated in this PR and I'm not sure whether I understand your position:

> Admission is wired for resources served by the server, e.g. APIServices for the aggregator. But admission is not active on any other HTTP endpoint like the proxy endpoints of the aggregator.

I do agree with this - I think my previous message was just an unnecessarily verbose way to try to express this message.

Basically, we just want to enable an admission webhook on service-catalog. Enabling it on the kubernetes apiserver did not enable it for service-catalog running under aggregation.

> I would propose: rebase to 1.10 and use the RecommendedOptions. Add your own admission controllers to the recommended list in the AdmissionOptions.

I am unfamiliar with Kubernetes codebase. I thought the change to use `--enable-admission-plugins` and `RecommendedOptions` will only change any custom apiserver that added options via `RecommendedOptions`?

From my understanding, the current thing that service-catalog does is to

1. Use the `Admission` options from apiserver already
2. Service Catalog installs it's own admission plugins (plugin/pkg/admission) into the AdmissionOptions in addition to Kubernetes ones.
3. It calls its own custom initializer which is required to call some internal interface methods instead of using the one that `Admission.ApplyTo` runs (service-catalog never calls ApplyTo). However, it incorrectly initializes Kubernetes admission plugins:
  * SetInternalServiceCatalogClientSet(internalclientset.Interface)
  * SetInternalServiceCatalogInformerFactory(informers.SharedInformerFactory)

The problem this PR is dealing with is (3) - the custom initializer that service-catalog ended up creating is not initializing the kubernetes ones. Will `RecommendedOptions` be enough - won't we also need to fork the initializer from 1.10 and add those custom init methods? Thanks for your help.

> This seems pretty huge. Every different apiserver has different admission controllers, and need to be manually synced up or manually configured each time. Even if we did provide this, how do we make sure people know about it to use it?

Hey @MHBauer - Sorry if I don't understand - I thought if a custom apiserver uses `RecommendedOptions` or `options.Admission`, they should all have the same default admission plugins provided by kubernetes apiserver? I'm basing my findings off [sample-apiserver](https://github.com/kubernetes/sample-apiserver) - I successfully ran "ValidatingAdmissionWebhook" without changing the code significantly.

It's only because service-catalog also requires `BrokerAuthSarCheck`, `ServiceBindingsLifecycle`, `ServicePlanChangeValidator` and `DefaultServicePlan` to be installed (`KubernetesNamespaceLifecycle` is just a copy-paste of the kubernetes `NamespaceLifecycle`)

Thanks for the review!

**Comment by sttts** [comment]
> added options via RecommendedOptions

Yes. And I hope that service catalog is using that. If not, it should. Is there any reason not use that?

> won't we also need to fork the initializer from 1.10 and add those custom init methods

No need to fork anything. Compare https://github.com/kubernetes/kubernetes/blob/a6c43c6a5ca7cc4449684d5e68d73773be91cd41/staging/src/k8s.io/sample-apiserver/pkg/cmd/server/start.go#L105 as an example of a custom initializer in an aggregated apiserver.

**Comment by ychen-atlassian** [comment]
Might be good to close this PR and reraise one to:

- Rebase on 1.10
- Use RecommendedOptions
- Fix bespoke options initializing code and use ApplyTo() instead (?)
- Use custom initializer code (https://github.com/kubernetes/kubernetes/blob/a6c43c6a5ca7cc4449684d5e68d73773be91cd41/staging/src/k8s.io/sample-apiserver/pkg/cmd/server/start.go#L105)

**Comment by pmorie** [comment]
@ychen-atlassian thanks a lot for this patch

> Might be good to close this PR and reraise one to:

I approve of what you wrote there.
